### ISSUE TITLE
[Feature Request] SWB upgrade in production

### ISSUE THREAD

**Issue by sxinger** [Missouri-BMI/service-workbench-on-aws#10]
blocked by https://github.com/orgs/Missouri-BMI/projects/9/views/46

- [x] We will need to fixed the following issues before upgrading to the new version with TRE:
      - s3-synchronizer issue #8 
      - study roleArn drift issue #9 
      - Implement auto stop on idle for EC2 instances [NextGen_Cloud_Architecture_AWS/issues/2](https://github.com/Missouri-BMI/NextGen_Cloud_Architecture_AWS/issues/2)
      - workspace termination issue due to SSM-related role drift #15 
      - document post-deployment manual configurations for special workspaces #19 
      
- [x] Upgrade scheduling
      - want to form a SWB user community on Teams (who should be the host?) [https://github.com/Missouri-BMI/service-workbench-on-aws/issues/20](https://github.com/Missouri-BMI/service-workbench-on-aws/issues/20)
      - we will send out first "downtime" email *1 week* prior to scheduled time to current user mail list to request them to: 
               - Step 1: save all of their data to a permanent location, D:/
               - Step 2: terminate their workspaces
               - Step 3: make sure to complete Step 1&2 prior to scheduled downtime, otherwise, they will risk losing their work
      - we will send out second "downtime" email on the day of upgrading as a reminder and let users know about the expected downtime duration
      - we will send out an "uptime" email once the upgrade is done with new features information
      - we won't be able to finalized the upgrade time until the pre-requisite issues are fixed. 

- [x] Upgrade management
       - fetch upstream https://github.com/awslabs/service-workbench-on-aws repo, once the new version is released. And perform upgrade on Sandbox or Application-Dev account and validate  #14 
       - after informing end-users with both "downtime" emails, we will perform upgrade on Application-Prod account

**Comment by fergusonsha-umkc** [comment]
Following SDLC best practices, once the anticipated features get merged into the main branch, I'll stage the updated release in our Sandbox account to allow for a baseline assessment before we reintroduce our customizations wherever still needed.  Then we will first upgrade Dev, and give it a full battery of tests of known issues to make sure they are all resolved and nothing new arises, before we initiate a scheduled maintenance window for upgrade of the Prod environment.

**Comment by fergusonsha-umkc** [comment]
Note from Travis, as a word of caution:
"Once you enable TRE in an environment, you cannot disable it.  If you ever decide that you no longer want or need TRE, you will have to completely delete the environment and reinstall it.  So please make sure you test it thoroughly before enabling it in your production environment. "

**Comment by fergusonsha-umkc** [comment]
Scheduling to begin deploying this starting with Sandbox, in January

**Comment by sxinger** [comment]
@fergusonsha-umkc @vasanthi014 @soliman455  I'm hoping that we will be able to have this upgrade done by March, is that possible? One major reason for this timeline is that we will need to go through annual security review with both CMS and University Info in May, and I think having SWB workspaces on private subnet would be an important security enhancement.  Also, Russ was hoping that we can have the web-based R workspace in place as well.

**Comment by fergusonsha-umkc** [comment]
We can make that a goal date yes, and I've updated my manager @magronef on this expectation, which will likely require my focus throughout February to achieve.  I'll have a better sense of whether this can be completed without needing to call on AWS technical assistance, after Sandbox is upgraded.  As TRE is a complete paradigm shift in how the environment will work, not just updating software compnents, it is a non-trivial upgrade times three, to work through, and I expect we will encounter unknowns that may delay the timeline.  We will, of course, try for the goal, but I want to set expectations in kind.

**Comment by fergusonsha-umkc** [comment]
I aim to start working on Sandbox upgrade and switch to TRE, this coming Monday.

**Comment by fergusonsha-umkc** [comment]
We've hit a snag with the upgrade, AppStream isn't available in our region yet.  We're in talks with Travis.

**Comment by fergusonsha-umkc** [comment]
I'm creating an S3 bucket to maintain the backups of the codebase archive for each Service Workbench instance, under each respective account.  This will let us tar up the directory off the Cloud9 workspaces to preserve our configuration and customizations between upgrades, and then delete the Cloud9 instances so they don't incur cost after they are needed.

**Comment by fergusonsha-umkc** [comment]
We are going to pursue getting into the private beta for AppStream in us-east-2 so we can use the Trusted Research Environment feature of the SWB upgrade.  Latest from Travis:

"Let me get back to the service team and ask what the process is for the private beta. It’s usually just a matter of knowing what account to add to an allow-list. I didn’t realize you had the NDA in place already. (Thanks, Matt!) I will get back with you soon."

Sandbox deployment, testing and tweaking, meanwhile, is underway with it turned off

**Comment by fergusonsha-umkc** [comment]
Update from Travis:
"Thanks, I’ll send this info to the service team.  I’ll let you know when it’s ready to move forward, or if they need anything else."

**Comment by fergusonsha-umkc** [comment]
Update from Travis:
"The team said they would update the allow-list this week, so I just sent them a request for an update.  I will let you know what I find out."

**Comment by fergusonsha-umkc** [comment]
Good news from Travis:
"I heard back from the team.  They have a process that runs once per week that updates a list of approved accounts, then applies that to the AppStream services.  We just missed the staging window last week.  But they advised that it will be applied by the end of this week, so Thursday or Friday.  I’ll share an update when I’m provided with one.  Apologies for the delay."

**Comment by fergusonsha-umkc** [comment]
Verified AppStream is available in Sandbox now, working on reinstalling the SWB update and importing the AppStream-enabled onboard-account.yaml

**Comment by fergusonsha-umkc** [comment]
Checked a 'git pull' in the SWB source tree and there is an update to v5.0.0 with multiple bug-fixes, so will start this Sandbox upgrade again from the top, with AppStream available now, to be sure no steps were missed.  The EMR AMI was also failing to build, though we haven't been using it, I don't like things failing :)

**Comment by sxinger** [comment]
So that your effort won't be wasted, I really need to think about a good use case to leverage our EMR workspace :) @fergusonsha-umkc

**Comment by fergusonsha-umkc** [comment]
environment-deploy.sh and build-images steps completed successfully, next tackling the stack update with onboard-account.cfn.yaml which sets up the AppStream pieces...

**Comment by fergusonsha-umkc** [comment]
... eventually found docs/docs/post_deployment/appstream.md which was not in the Installation Guide PDF

**Comment by fergusonsha-umkc** [comment]
Making more progress, got the AppStream Image Builder kicked off now, using VPC and subnet fixed values.

**Comment by fergusonsha-umkc** [comment]
Soooo close!  AppStream pieces are all built now, but running into some difficulty with launching workspaces still, doing a fresh account onboard to see if it fixes things, as it seems to be using an old template still.

**Comment by fergusonsha-umkc** [comment]
Important step:
Once prerequisites are complete and the CloudFormation stack has been created, go to AppStream on the AWS console of the hosting account. Go to **Fleet** and then choose the newly created fleet. Choose **Action**>**Start** to start the fleet.  This step must be completed for any researcher workspace to be successfully launched within the hosting account.

**Comment by fergusonsha-umkc** [comment]
Please refer to this for latest progress, nearing release of 5.2.2 non-TRE upgrade in October ... https://github.com/Missouri-BMI/NextGen_Cloud_Architecture_AWS/issues/41 (not sure how we ended up with 2 of these entries to track everything in.

**Comment by fergusonsha-umkc** [comment]
Making a note of it here, version 5.2.3 was released (we are doing the Big Leap upgrade to 5.2.2 still) which addresses Node.JS 12 function deprecation (not near-term and will all still work).  So I propose using the v5.2.2 to v5.2.3 upgrade _after_ we get to v5.2.2 in Production, to prototype and document a more "routine" minor-version upgrade cycle as they make those available.  Nothing about the current leap from 3.1 to 5.2.2 is minor or routine :wink:

**Comment by fergusonsha-umkc** [comment]
Go me, I made an issue for that:  https://github.com/Missouri-BMI/service-workbench-on-aws/issues/32
